runtime/autoload/help.vim

```diff
@@ -0,0 +1,38 @@
+function! s:synnames(...) abort
+  if a:0
+    let [line, col] = [a:1, a:2]
+  else
+    let [line, col] = [line('.'), col('.')]
+  endif
+  return reverse(map(synstack(line, col), 'synIDattr(v:val,"name")'))
+endfunction
+
+function! HelpTopic()
+  let col = col('.') - 1
+  while col && getline('.')[col] =~# '\k'
+    let col -= 1
+  endwhile
+  let pre = col == 0 ? '' : getline('.')[0 : col]
+  let col = col('.') - 1
+  while col && getline('.')[col] =~# '\k'
+    let col += 1
+  endwhile
+  let post = getline('.')[col : -1]
+  let syn = get(synnames(), 0, '')
+  let cword = expand('<cword>')
+  if syn ==# 'vimFuncName'
+    return cword.'()'
+  elseif syn ==# 'vimOption'
```

Comment:
Topic for options: if current word is preceded with `&`, `&l:` or `&g:` or surrounded by `'`. Also if some preceding text contains `set`, `setl[ocal]` or `setg[lobal]`. Note: only in the latter case it *sometimes* uses `vimOption` syntax group (sometimes: because it may be `execute 'set rtp='.…). E.g. for `&syntax` it will *not* guess that it needs to open `'syntax'`.